core/store/src/trie/mem/resharding.rs

```diff
@@ -59,110 +87,166 @@ impl<'a, M: ArenaMemory> MemTrieUpdate<'a, M> {
     /// retain operation.
     fn retain_multi_range(mut self, intervals: &[Range<Vec<u8>>]) -> (TrieChanges, TrieAccesses) {
         debug_assert!(intervals.iter().all(|range| range.start < range.end));
-        let intervals_nibbles = intervals
-            .iter()
-            .map(|range| {
-                NibbleSlice::new(&range.start).iter().collect_vec()
-                    ..NibbleSlice::new(&range.end).iter().collect_vec()
-            })
-            .collect_vec();
+        let intervals_nibbles = intervals_to_nibbles(intervals);
 
         // TODO(#12074): consider handling the case when no changes are made.
         // TODO(#12074): restore proof as well.
-        self.retain_multi_range_recursive(0, vec![], &intervals_nibbles);
+        self.generic_retain_multi_range_recursive(0, vec![], &intervals_nibbles).unwrap();
         self.to_trie_changes()
     }
+}
+
+impl Trie {
+    pub fn retain_split_shard(
+        &self,
+        boundary_account: &AccountId,
+        retain_mode: RetainMode,
+    ) -> Result<StateRoot, StorageError> {
+        let intervals = boundary_account_to_intervals(boundary_account, retain_mode);
+        self.retain_multi_range(&intervals)
+    }
 
+    fn retain_multi_range(&self, intervals: &[Range<Vec<u8>>]) -> Result<StateRoot, StorageError> {
+        debug_assert!(intervals.iter().all(|range| range.start < range.end));
+        let intervals_nibbles = intervals_to_nibbles(intervals);
+
+        let mut memory = NodesStorage::new(&self);
+        let root_node = self.move_node_to_mutable(&mut memory, &self.root)?;
```

Comment:
I do not see docs for `move_node_to_mutable` and `flatten_nodes`, but do I get it correctly that `move_node_to_mutable` "pops" root from the trie, then `generic_retain_multi_range_recursive` leaves only nodes we want, and `flatten_nodes` rebuild the trie ?

Reply:
Correct. This pattern is used for processing insertions and deletions as well.